providers/in-process-flagd/provider.go

```diff
@@ -0,0 +1,644 @@
+				p.logger.Error(
+					fmt.Sprintf(
+						"Invalid env config for %s provided, using default value: %s",
+						flagdSyncRetryIntervalEnvironmentVariableName,
+						defaultBackoffDuration.String(),
+					),
+				)
+			} else {
+				p.connectionInfo.backoffDuration = maxSyncRetryInterval
+			}
+		}
+
+		sourceURI := os.Getenv(flagdSourceURIEnvironmentVariableName)
+		sourceProvider := os.Getenv(flagdSourceProviderEnvironmentVariableName)
+		selector := os.Getenv(flagdSourceSelectorEnvironmentVariableName)
+
+		if sourceURI != "" && sourceProvider != "" {
+			p.providerConfiguration.SourceConfig = runtime.SourceConfig{
+				URI:      sourceURI,
+				Provider: sourceProvider,
+				Selector: selector,
+				CertPath: p.providerConfiguration.CertificatePath,
+				TLS:      p.providerConfiguration.TLSEnabled,
+			}
+		}
+	}
+}
+
+// WithSourceURI sets the URI of the sync source
+func WithSourceURI(uri string) ProviderOption {
+	return func(p *Provider) {
+		p.providerConfiguration.SourceConfig.URI = uri
+	}
+}
+
+func WithSourceType(providerType ProviderType) ProviderOption {
+	return func(p *Provider) {
+		p.providerConfiguration.SourceConfig.Provider = string(providerType)
+	}
+}
+
+// WithContext supplies the given context to the event stream. Not to be confused with the context used in individual
+// flag evaluation requests.
+func WithContext(ctx context.Context) ProviderOption {
+	return func(p *Provider) {
+		p.ctx = ctx
+	}
+}
+
+// WithSyncStreamConnectionMaxAttempts sets the maximum number of attempts to connect to flagd's event stream.
+// On successful connection the attempts are reset.
+func WithSyncStreamConnectionMaxAttempts(i int) ProviderOption {
+	return func(p *Provider) {
+		p.connectionInfo.maxSyncRetries = i
+	}
+}
+
+// WithSyncStreamConnectionBackoff sets the backoff duration between reattempts of connecting to the sync source.
+func WithSyncStreamConnectionBackoff(duration time.Duration) ProviderOption {
+	return func(p *Provider) {
+		p.connectionInfo.backoffDuration = duration
+	}
+}
+
+// WithLogger sets the logger used by the provider.
+func WithLogger(l *logger.Logger) ProviderOption {
+	return func(p *Provider) {
+		p.logger = l
+	}
+}
+
+// WithTLS enables TLS. If certPath is not given, system certs are used.
+func WithTLS(certPath string) ProviderOption {
+	return func(p *Provider) {
+		p.providerConfiguration.TLSEnabled = true
+		p.providerConfiguration.CertificatePath = certPath
+	}
+}
+
+// WithSelector sets the selector for the sync source
+func WithSelector(selector string) ProviderOption {
+	return func(p *Provider) {
+		p.providerConfiguration.SourceConfig.Selector = selector
+	}
+}
+
+// Status returns the current provider status
+func (p *Provider) Status() of.State {
+	switch p.connectionInfo.state {
+	case stateNotReady:
+		return of.NotReadyState
+	case stateReady:
+		return of.ReadyState
+	case stateStale:
+		// currently there is no STALE state in the of API, so we also return ERROR for STALE
```

Comment:
This is a good point... I wonder if we should delay the actual release of this provider until spec `v0.7.0` is implemented in Go (there wouldn't be too much work required).

Reply:
cc @Kavindu-Dodan @mowies  - we may want to consider something similar for Java/JS, neither of them have `STALE` yet (though there's a [draft PR](https://github.com/open-feature/js-sdk/pull/541) for JS)

Reply:
Given that this is part of the spec [^1], state `STALE` should be added to SDK. 

[^1]: https://openfeature.dev/specification/types#provider-events

Reply:
ohh my bad, provider event stale is not yet reflected in provider state. This PR need to be merged for this [^1]

[^1]: https://github.com/open-feature/spec/pull/200

Reply:
Stale state actually is in the spec `main`, just not release or implemented in very many SDKs: https://github.com/open-feature/spec/blob/main/specification/types.md#provider-status

This will come with spec `0.7.0` release. I'll try to do that tomorrow: https://github.com/open-feature/spec/issues/202